File: Week_06/java/src/_91_DecodeWays.java
/**
 * 第一遍：2020/07/26周六 ✅
 * 第一遍：2020/07/27周日 ✅
 * 第二遍：2020/06/15周一
 * 第三遍：2020/06/21周日
 * 第四遍：2020/07/05周日
 */
class _91_DecodeWays {
    public int numDecodings(String s) {
        //f(i)前i个字符解码方法的总数
        //if (s[i-1] == '0') 
        //      if (s[i - 2] == '1' || s[i - 2] == '2') f(i) = f(i - 2) else return 0
        //else if (s[i - 2] == '1' || (s[i - 2] == '2' && s[i - 1] <= '6') 
        //      f(i) = f(i - 1) + f(i - 2)
        //else f(i) = f(i - 1)
        //f(0) = 0 f(1) = 1 i in [2, s.length()]
        if (s == null || s.length() == 0 || s.charAt(0) == '0') return 0;
        int length = s.length();
        int leftPreceding = 1;
        int left = 1;
        int dp = 1;
        for (int i = 2; i <= length; i++) {
            if (s.charAt(i - 1) == '0') {
                if (s.charAt(i - 2) == '1' || s.charAt(i - 2) == '2') {
                    dp = leftPreceding;
                } else {
                    return 0;
                }
            } else {
                if (s.charAt(i - 2) == '1' || (s.charAt(i - 2) == '2' && s.charAt(i - 1) <= '6')) {
                    dp = left + leftPreceding;
                } else {
                    dp = left;
                }
            }
            leftPreceding = left;
            left = dp;
        }
        return dp;
    }
}

File: Week_06/kotlin/src/_91_DecodeWays.kt
fun numDecodings(s: String): Int {
    //f(i)前i个字符的解码方法数
    //if (s[i - 1] == '0') if (s[i - 2] == '1' || s[i - 2] == '2') f(i) = f(i - 2) else f(i) =  0
    //else if (s[i - 2] == '1' || (s[i - 2] == '2' && s[i - 1] <= '6')) f(i) = f(i - 1) + f(i - 2) else f(i) = f(i - 1)
    //i in [1, s.size] f(0) = 1 f(1) = 1
    //if (s[0] == '0') return 0
    if (s.isEmpty() || s[0] == '0') return 0
    var leftPreceding = 1
    var left = 1
    for (i in 2..s.length) {
        val first = s[i - 1] - '0'
        val second = (s[i - 2] - '0') * 10 + first
        var dp = 0
        if (first in 1..9) {
            dp = left
        }
        if (second in 10..26) {
            dp += leftPreceding
        }
        leftPreceding = left
        left = dp
    }
    return left;
}